interfaces/apparmor/backend_test.go

```diff
@@ -0,0 +1,374 @@
+		s.removeSnap(c, snapInfo)
+	}
+}
+
+func (s *backendSuite) TestRealDefaultTemplateIsNormallyUsed(c *C) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(sambaYamlV1))
+	c.Assert(err, IsNil)
+	// NOTE: we don't call apparmor.MockTemplate()
+	err = s.backend.Configure(snapInfo, false, s.repo)
+	c.Assert(err, IsNil)
+	profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.smbd")
+	data, err := ioutil.ReadFile(profile)
+	c.Assert(err, IsNil)
+	for _, line := range []string{
+		// NOTE: a few randomly picked lines from the real profile.  Comments
+		// and empty lines are avoided as those can be discarded in the future.
+		"#include <tunables/global>\n",
+		"/tmp/   r,\n",
+		"/sys/class/ r,\n",
+	} {
+		c.Assert(string(data), testutil.Contains, line)
+	}
+}
+
+func (s *backendSuite) TestCustomTemplateUsedOnRequest(c *C) {
+	s.backend.CustomTemplate = `
+# Description: Custom template for testing
+###VAR###
+
+###PROFILEATTACH### (attach_disconnected) {
+	FOO
+}
+`
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(sambaYamlV1))
+	c.Assert(err, IsNil)
+	err = s.backend.Configure(snapInfo, false, s.repo)
+	c.Assert(err, IsNil)
+	profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.smbd")
+	data, err := ioutil.ReadFile(profile)
+	c.Assert(err, IsNil)
+	// Our custom template was used
+	c.Assert(string(data), testutil.Contains, "FOO")
+	// Custom profile can rely on legacy variables
+	for _, legacyVarName := range []string{
+		"APP_APPNAME", "APP_ID_DBUS", "APP_PKGNAME_DBUS",
+		"APP_PKGNAME", "APP_VERSION", "INSTALL_DIR",
+	} {
+		c.Assert(string(data), testutil.Contains, fmt.Sprintf("@{%s}=", legacyVarName))
+	}
+}
+
+type combineSnippetsScenario struct {
+	developerMode bool
+	snippet       string
+	content       string
+}
+
+var combineSnippetsScenarios = []combineSnippetsScenario{{
+	content: `
+@{APP_APPNAME}="smbd"
+@{APP_ID_DBUS}="samba_2eacme_5fsmbd_5f1"
+@{APP_PKGNAME_DBUS}="samba_2eacme"
+@{APP_PKGNAME}="samba.acme"
+@{APP_VERSION}="1"
+@{INSTALL_DIR}="{/snaps,/gadget}"
+profile "snap.samba.smbd" (attach_disconnected) {
+}
+`,
+}, {
+	snippet: "snippet",
+	content: `
+@{APP_APPNAME}="smbd"
+@{APP_ID_DBUS}="samba_2eacme_5fsmbd_5f1"
+@{APP_PKGNAME_DBUS}="samba_2eacme"
+@{APP_PKGNAME}="samba.acme"
+@{APP_VERSION}="1"
+@{INSTALL_DIR}="{/snaps,/gadget}"
+profile "snap.samba.smbd" (attach_disconnected) {
+snippet
+}
+`,
+}, {
+	developerMode: true,
+	content: `
+@{APP_APPNAME}="smbd"
+@{APP_ID_DBUS}="samba_2eacme_5fsmbd_5f1"
+@{APP_PKGNAME_DBUS}="samba_2eacme"
+@{APP_PKGNAME}="samba.acme"
+@{APP_VERSION}="1"
+@{INSTALL_DIR}="{/snaps,/gadget}"
+profile "snap.samba.smbd" (attach_disconnected,complain) {
+}
+`,
+}, {
+	developerMode: true,
+	snippet:       "snippet",
+	content: `
+@{APP_APPNAME}="smbd"
+@{APP_ID_DBUS}="samba_2eacme_5fsmbd_5f1"
+@{APP_PKGNAME_DBUS}="samba_2eacme"
+@{APP_PKGNAME}="samba.acme"
+@{APP_VERSION}="1"
+@{INSTALL_DIR}="{/snaps,/gadget}"
+profile "snap.samba.smbd" (attach_disconnected,complain) {
+snippet
```

Comment:
Again, can we please kill the duplication in these snippets?  The actual relevant difference in them is *5 lines*, total! We don't need all 34 here. For instance:
```
content: commonPrefix + `
...
```